File: frontend/lib/types/certification.ts
export interface Microtopic {
  id: string;
  subject: string;
  topic: string;
  label: string;
}

export type MessageRole = "tutor" | "student";

export interface ChatMessage {
  role: MessageRole;
  content: string;
  created_at: string;
}

export interface StartCertificationResponse {
  certification_id: string;
  student_facing_blurb: string;
  opening_message: string;
  min_turns_to_end: number;
  max_turns: number;
}

export interface SendMessageResponse {
  reply: string;
  turn_count: number;
}

export interface RubricScore {
  subject_knowledge: number;
  subject_knowledge_rationale: string;
  instructional_quality: number;
  instructional_quality_rationale: string;
  pedagogical_adaptability: number;
  pedagogical_adaptability_rationale: string;
  organization: number;
  organization_rationale: string;
  total_score: number;
  passed: boolean;
}

export interface EndCertificationResponse {
  score: RubricScore;
}

export interface CertificationState {
  certification_id: string;
  status: "in_progress" | "completed";
  transcript: ChatMessage[];
  tutor_turn_count: number;
  score: RubricScore | null;
  min_turns_to_end: number;
  max_turns: number;
}

File: backend/app/models/certification.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.models.message import Message
from app.models.rubric import RubricScore


class StartCertificationRequest(BaseModel):
    microtopic_id: UUID


class StartCertificationResponse(BaseModel):
    certification_id: UUID
    student_facing_blurb: str
    opening_message: str
    min_turns_to_end: int
    max_turns: int


class SendMessageRequest(BaseModel):
    content: str = Field(min_length=1, max_length=4000)

    @field_validator("content")
    @classmethod
    def not_blank(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("content cannot be blank")
        return stripped


class SendMessageResponse(BaseModel):
    reply: str
    turn_count: int


class EndCertificationResponse(BaseModel):
    score: RubricScore


class CertificationStateResponse(BaseModel):
    certification_id: UUID
    status: Literal["in_progress", "completed"]
    transcript: list[Message]
    tutor_turn_count: int
    score: RubricScore | None = None
    min_turns_to_end: int
    max_turns: int
